### PR TITLE
Ruby Gem

### PR DESCRIPTION
**Pull request by nihalgonsalves**
Features added

- Load details from the API at ifsc.razorpay.com (see [README.md](https://github.com/nihalgonsalves/ifsc/blob/9846a04184090610f920c523f5c99cc86064b7ba/src/ruby/README.md) for usage)
- Test coverage for all added features

Fixes

- Use memoized JSON parsing instead of constants
- Fix string branch code lookups
- Remove usage of `and` instead of `&&`

Changes

- Set `lookup_` methods to private
- Use `<< self` for the block of class methods
- Use `#valid?` instead of `#validate` for semantics
- `Razorpay::IFSC` –> `Razorpay::IFSC::Code` to allow for errors within the `IFSC` namespace
- Add `#validate!` and `Razorpay::IFSC::InvalidCodeError`
- Use `rspec` instead of `minitest` to mock better

Todo

- [ ] Test Gem publishing and get the JSON file into the final gem bundle

### CODE REVIEW

File: src/ruby/ifsc.gemspec [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+Gem::Specification.new do |s|
+  s.name        = 'ifsc'
+  s.version     = '0.0.1'
+  s.date        = '2016-05-13'
+  s.summary     = "IFSC code database to help you validate IFSC codes"
+  s.description = "A simple gem by @razorpay to help you validate your IFSC codes. IFSC codes are bank codes within India"
+  s.authors     = ['Abhay Rana']
+  s.email       = ['nemo@razorpay.com']
+  s.files       = ['lib/IFSC.rb', 'spec/ifsc_spec.rb', 'Rakefile', 'Gemfile', 'ifsc.gemspec']
```
Comment (captn3m0) [inline src/ruby/ifsc.gemspec]: lower case IFSC.rb

File: src/ruby/README.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+code.state
+# => "HARYANA"
+```
+
+### Error handling
+
+```rb
+# all these `Razorpay::IFSC::InvalidCodeError` for an invalid code
+Razorpay::IFSC::Code.validate! '...'
+Razorpay::IFSC::Code.find '...'
```
Comment (captn3m0) [inline src/ruby/README.md]: s/Code/Client just for consistency? (the PHP one uses `Razorpay\IFSC\IFSC` and `Razorpay\IFSC\Client`.

File: src/ruby/lib/ifsc.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        def data
+          @data ||= JSON.load(File.read(File.join(__dir__, '../../IFSC.json')))
+        end
+
+        def lookup_numeric(list, branch_code)
+          branch_code = branch_code.to_i
+
+          return true if list.include? branch_code
+
+          lookup_ranges list, branch_code
```
Comment (captn3m0) [inline src/ruby/lib/ifsc.rb]: lookup_ranges is no longer supported because the underlying IFSC.json now holds everything

**Comment by captn3m0** [comment]
@nihalgonsalves 

Thanks a lot for this. The constants are quite helpful when you are dealing with bank-specific code (which is true in our case). We can leave them for the first release while we build parity with the PHP version. Other differences I notices in this quick review:

1. sublet branches are not supported (Which is fine for a first release)
2. lookup_ranges is not used in PHP anymore and the json file doesn't support it either, so it is redundant.
3. IFSC-list.json and IFSC.json are confusing. The package only ships with one `IFSC.json` which you can consult against the master
4. banknames.json is also not used. This is helpful because it standardizes all the bank names and they are maintained for display niceness.

**Comment by nihalgonsalves** [comment]
@captn3m0 Thanks for the review!

The IFSC-list and IFSC json files are something from your initial commit, I didn't want to remove them in case you needed them for something. 

- [Compare initial ruby commit + merged master to the current code](https://github.com/nihalgonsalves/ifsc/compare/a6711ba576128a954d69ce3dcc9eecfabb7856f0...ruby)
- Your initial code: [Compare razorpay/master ... razorpay/ruby](https://github.com/razorpay/ifsc/compare/master...ruby)

Do you have a specification for the ifsc/sublet formats? I actually looked at the node client to fix this one, so I didn't come across the sublet logic. Looking briefly at the PHP code, it looks like it should be fairly simple to add. It would probably be a good idea to add a kind of specification for the validation, data files and other logic in pseudocode or just plain English. (This would also make it easier to develop new clients).

`banknames.json` - Where would this be used? I currently grab the data from your ifsc.razorpay.com API. Do you mean that the returned bank names should come from this file instead?

I'll make the rest of the changes later when I have some free time later.

**Comment by captn3m0** [comment]
Specification sounds like a nice idea. I'll write down a "Packaging Guide" that covers all of these points to make it easier for you and anyone else by today.

The bank_name support can be dropped now, because the API itself returns the correct bankname now. (Corresponding PR: https://github.com/razorpay/ifsc-api/pull/5).

There is still a IFSC\getBankName method that takes in the IFSC code and returns the name completely offline, which I think is nice to do. I also just noticed that I haven't yet cleared the range lookup code in PHP either, so I'll try to clean that up.

**Comment by nihalgonsalves** [comment]
Sounds good. I can add the offline bank name method.

I'm not entirely familiar with how often the IFSC database is updated - but would it be a concern if I include this gem (expecting it to always work) and then their application rejects valid IFSC codes when the data gets updated?

In that case, maybe we could provide an online-only validation method and/or fetch and memoize the IFSC.json file from the server within the class, instead of from disk?

**Comment by captn3m0** [comment]
We run this in production, and the idea is to keep doing minor/patch updates everytime the dataset is updated (which is once a month usually, depending on RBI). You can check the releases page for more details on frequency etc.

The newer IFSC codes that are added usually take some time before they start getting used, so the expectation is to keep the gem updated. (I'd recommend subscribing the releases RSS feed).

We can clarify this in the README itself, something like: Use IFSC::validate() for offline validation, but you'll have to keep the package updated. Alternatively, you can use the IFSC\Client::lookupIFSC method which makes a web request to the IFSC API for validation instead.

**Comment by nihalgonsalves** [comment]
- Class name changed to be consistent with the PHP library
- Offline bank name lookup `obj.get_bank_name` / `class.bank_name_for?(code)` that uses sublet lookups too.
- Online only validation with `obj.get` and `class.find(code)`. Throws `InvalidCodeError` on 404 and `ServerError` on any other API related error.

File: src/ruby/lib/ifsc.rb (hunk 3 above)
Comment (nihalgonsalves) [inline src/ruby/lib/ifsc.rb]: with the latest `IFSC.json` from master, it still looks like `lookup_ranges` need to be used. The tests fail if I remove the `lookup_ranges` check. Am I missing something?

Comment (nihalgonsalves) [inline src/ruby/lib/ifsc.rb]: Hm, never mind, I was mistaken.

**Comment by captn3m0** [comment]
@nihalgonsalves Can you tick the button that grants collaborators write access. @harman28 is doing a review as well. We're looking to release the package soon, and it would be easier if we have write access.

You can do a review once any (if any) changes go through before we do the release.

**Comment by nihalgonsalves** [comment]
@captn3m0 @harman28 

You should have access. Its been ticked since I created this PR.

**Comment by captn3m0** [comment]
@nihalgonsalves Merged this. Will be pushing a ifsc gem soon

**Comment by captn3m0** [comment]
@nihalgonsalves After some restructuring (#34) , pushed out a release today: https://rubygems.org/gems/ifsc

**Comment by nihalgonsalves** [comment]
@captn3m0 Thanks for the update, nice to see this published!

P.s. do you think you could add me as collaborator/author on the Gemspec? ;)

Cheers!

**Pull request merged**